File: electron/sqlite/queries/audit.js
/**
 * 审计日志 SQLite 查询模块
 *
 * 实现 audit_logs 表的 CRUD 操作。
 *
 * 注意字段映射差异：
 * - JSON `timestamp`  ↔ SQLite `created_at`
 * - JSON `username`   ↔ SQLite `user_name`
 * - JSON `resource`   ↔ SQLite `resource_type`
 * - JSON `description` ↔ SQLite `details`（描述性文本存 details 列）
 * - JSON `resourceName` → SQLite 无此列（查询时用子查询模拟）
 * - JSON `ip`         ↔ SQLite `ip_address`
 */
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import log from 'electron-log';
import { tryGetSqlite, toSqliteValue, useSqliteRead } from './helpers';
// ═══════════════════════════════════════════════════════════════════════════════
// 字段映射
// ═══════════════════════════════════════════════════════════════════════════════
/**
 * 将 JSON 审计日志转为 SQLite INSERT 参数
 *
 * JSON AuditLog 字段 → SQLite audit_logs 列：
 *   id          → id (TEXT, 格式 "log_123_abc")
 *   timestamp   → created_at
 *   userId      → user_id
 *   username    → user_name
 *   action      → action
 *   resource    → resource_type
 *   resourceId  → resource_id
 *   level       → level
 *   description → details
 *   ip          → ip_address
 */
function auditLogToParams(auditLog) {
    return [
        toSqliteValue(auditLog.id), // id (TEXT)
        toSqliteValue(auditLog.action), // action
        toSqliteValue(auditLog.level || 'info'), // level
        toSqliteValue(auditLog.userId), // user_id
        toSqliteValue(auditLog.username), // user_name
        toSqliteValue(auditLog.resource), // resource_type
        toSqliteValue(auditLog.resourceId), // resource_id
        toSqliteValue(auditLog.description), // details
        toSqliteValue(auditLog.ip), // ip_address
        toSqliteValue(auditLog.timestamp || new Date().toISOString()), // created_at
    ];
}
/** INSERT 语句 */
var INSERT_SQL = "\n  INSERT OR REPLACE INTO audit_logs\n    (id, action, level, user_id, user_name, resource_type, resource_id, details, ip_address, created_at)\n  VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)\n";
/**
 * 将 SQLite 行转为前端期望的 AuditLog 格式
 */
function rowToAuditLog(row) {
    return {
        id: row.id,
        action: row.action,
        level: row.level || 'info',
        userId: row.user_id,
        username: row.user_name,
        resource: row.resource_type,
        resourceId: row.resource_id,
        description: row.details,
        ip: row.ip_address,
        timestamp: row.created_at,
        // resourceName 在 SQLite 中没有对应列，设为 null
        resourceName: null,
        // details 在 SQLite 中复用 details 列（description），
        // 原始 details 对象在 JSON 迁移时已丢失（前端 audit.ts 会剥离 details 再存 JSON）
    };
}
// ═══════════════════════════════════════════════════════════════════════════════
// 写操作
// ═══════════════════════════════════════════════════════════════════════════════
/**
 * 记录审计日志（SQLite 版）
 */
export function logAudit(auditLog) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var params = auditLogToParams(auditLog);
        (_a = sqlite.prepare(INSERT_SQL)).run.apply(_a, params);
        return true;
    }
    catch (err) {
        log.error('[SQLite] audit.log error:', err);
        return false;
    }
}
/**
 * 清理旧日志（SQLite 版）
 * @returns 删除的行数
 */
export function clearLogs(daysToKeep) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return 0;
    try {
        var cutoffDate = new Date(Date.now() - daysToKeep * 24 * 60 * 60 * 1000).toISOString();
        var result = sqlite.prepare('DELETE FROM audit_logs WHERE created_at < ?').run(cutoffDate);
        return result.changes;
    }
    catch (err) {
        log.error('[SQLite] audit.clear error:', err);
        return 0;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// 读操作
// ═══════════════════════════════════════════════════════════════════════════════
/**
 * 查询审计日志（SQLite 版）
 * 返回格式与 JSON 版 audit:query 一致
 */
export function queryLogs(query) {
    var _a, _b;
    if (!useSqliteRead())
        return null;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        var conditions = [];
        var params = [];
        // 日期筛选
        if (query.startDate) {
            conditions.push('created_at >= ?');
            params.push(query.startDate);
        }
        if (query.endDate) {
            var endDate = new Date(query.endDate);
            endDate.setHours(23, 59, 59, 999);
            conditions.push('created_at <= ?');
            params.push(endDate.toISOString());
        }
        // 操作类型
        if (query.action) {
            conditions.push('action = ?');
            params.push(query.action);
        }
        // 资源类型（JSON 的 resource 对应 SQLite 的 resource_type）
        if (query.resource) {
            conditions.push('resource_type = ?');
            params.push(query.resource);
        }
        // 级别
        if (query.level) {
            conditions.push('level = ?');
            params.push(query.level);
        }
        // 关键词搜索（user_name / details / resource_id）
        if (query.keyword) {
            conditions.push('(user_name LIKE ? OR details LIKE ? OR resource_id LIKE ?)');
            var kw = "%".concat(query.keyword.toLowerCase(), "%");
            params.push(kw, kw, kw);
        }
        var whereClause = conditions.length > 0 ? "WHERE ".concat(conditions.join(' AND ')) : '';
        // 总数
        var countRow = (_a = sqlite.prepare("SELECT COUNT(*) as count FROM audit_logs ".concat(whereClause)))
            .get.apply(_a, params);
        var total = countRow.count;
        // 分页
        var page = query.page || 1;
        var pageSize = query.pageSize || 20;
        var totalPages = Math.ceil(total / pageSize);
        var offset = (page - 1) * pageSize;
        // 查询
        var rows = (_b = sqlite.prepare("SELECT * FROM audit_logs ".concat(whereClause, " ORDER BY created_at DESC LIMIT ? OFFSET ?"))).all.apply(_b, __spreadArray(__spreadArray([], params, false), [pageSize, offset], false));
        var items = rows.map(rowToAuditLog);
        return { items: items, total: total, page: page, pageSize: pageSize, totalPages: totalPages };
    }
    catch (err) {
        log.error('[SQLite] audit.query error:', err);
        return null;
    }
}
/**
 * 获取审计统计（SQLite 版）
 */
export function getStats(days) {
    var _a, _b, _c, _d, _e;
    if (!useSqliteRead())
        return null;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        var dateFilter = '';
        var params = [];
        if (days) {
            var cutoffDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString();
            dateFilter = 'WHERE created_at >= ?';
            params.push(cutoffDate);
        }
        // 总数
        var countRow = (_a = sqlite.prepare("SELECT COUNT(*) as count FROM audit_logs ".concat(dateFilter)))
            .get.apply(_a, params);
        var totalCount = countRow.count;
        // 今日操作数
        var today = new Date().toISOString().split('T')[0];
        var todayParams = __spreadArray([], params, true);
        var todayWhere = dateFilter
            ? "".concat(dateFilter, " AND created_at >= ?")
            : 'WHERE created_at >= ?';
        todayParams.push(today);
        var todayRow = (_b = sqlite.prepare("SELECT COUNT(*) as count FROM audit_logs ".concat(todayWhere)))
            .get.apply(_b, todayParams);
        var todayCount = todayRow.count;
        // 操作类型分布
        var actionRows = (_c = sqlite.prepare("SELECT action, COUNT(*) as count FROM audit_logs ".concat(dateFilter, " GROUP BY action"))).all.apply(_c, params);
        var actionCounts = {};
        for (var _i = 0, actionRows_1 = actionRows; _i < actionRows_1.length; _i++) {
            var row = actionRows_1[_i];
            actionCounts[row.action] = row.count;
        }
        // 资源类型分布
        var resourceRows = (_d = sqlite.prepare("SELECT resource_type, COUNT(*) as count FROM audit_logs ".concat(dateFilter, " GROUP BY resource_type"))).all.apply(_d, params);
        var resourceCounts = {};
        for (var _f = 0, resourceRows_1 = resourceRows; _f < resourceRows_1.length; _f++) {
            var row = resourceRows_1[_f];
            resourceCounts[row.resource_type] = row.count;
        }
        // 活跃用户
        var userRows = (_e = sqlite.prepare("SELECT user_id, user_name, COUNT(*) as count FROM audit_logs ".concat(dateFilter, " GROUP BY user_id, user_name ORDER BY count DESC LIMIT 10"))).all.apply(_e, params);
        var topUsers = userRows.map(function (row) { return ({
            userId: row.user_id || 'unknown',
            username: row.user_name || 'unknown',
            count: row.count,
        }); });
        return { totalCount: totalCount, todayCount: todayCount, actionCounts: actionCounts, resourceCounts: resourceCounts, topUsers: topUsers };
    }
    catch (err) {
        log.error('[SQLite] audit.stats error:', err);
        return null;
    }
}

File: src-tauri/src/commands/audit.rs
/**
 * 审计日志命令
 *
 * 对应 Electron 版本的 audit.ts
 * 包含日志记录、查询、统计、清理
 */

use crate::db::AppState;
use crate::error::{AppError, AppResult};
use rusqlite::params;
use serde::{Deserialize, Serialize};
use tauri::{command, State};

// ============ 类型定义 ============

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AuditLogEntry {
    pub id: Option<String>, // 前端传字符串 ID，忽略即可
    pub user_id: Option<String>,
    pub username: Option<String>,
    pub action: String,
    pub resource: Option<String>, // 前端用 resource 字段
    pub resource_id: Option<String>,
    pub resource_name: Option<String>,
    pub level: Option<String>,
    pub description: Option<String>,
    pub details: Option<String>,
    pub ip_address: Option<String>,
    pub timestamp: Option<String>, // 前端用 timestamp 字段
    pub created_at: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct AuditQuery {
    pub start_date: Option<String>,
    pub end_date: Option<String>,
    pub action: Option<String>,
    pub resource: Option<String>,
    pub level: Option<String>,
    pub keyword: Option<String>,
    pub page: Option<i64>,
    pub page_size: Option<i64>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AuditQueryResult {
    pub items: Vec<AuditLogEntry>,
    pub total: i64,
    pub page: i64,
    pub page_size: i64,
    pub total_pages: i64,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AuditStats {
    pub total_count: i64,
    pub today_count: i64,
    pub action_counts: std::collections::HashMap<String, i64>,
    pub resource_counts: std::collections::HashMap<String, i64>,
    pub top_users: Vec<TopUser>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TopUser {
    pub user_id: String,
    pub username: String,
    pub count: i64,
}

// ============ 命令实现 ============

/// 记录审计日志
#[command]
pub fn audit_log(state: State<'_, AppState>, entry: AuditLogEntry) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;
    let now = entry
        .created_at
        .unwrap_or_else(|| chrono::Local::now().format("%Y-%m-%d %H:%M:%S").to_string());

    db.execute(
        "INSERT INTO audit_logs (action, level, user_id, user_name, resource_type, resource_id, details, ip_address, created_at)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)",
        params![
            entry.action,
            entry.level.unwrap_or_else(|| "info".to_string()),
            entry.user_id.unwrap_or_default(),
            entry.username.unwrap_or_default(),
            entry.resource.unwrap_or_default(),
            entry.resource_id.unwrap_or_default(),
            entry.details.or(entry.description).unwrap_or_default(),
            entry.ip_address.unwrap_or_default(),
            entry.timestamp.unwrap_or(now),
        ],
    )
    .map_err(|e| AppError::Database(e.to_string()))?;

    // 保持上限 10000 条（与 Electron 版本一致）
    let _ = db.execute(
        "DELETE FROM audit_logs WHERE id NOT IN (SELECT id FROM audit_logs ORDER BY created_at DESC LIMIT 10000)",
        [],
    );

    Ok(())
}

/// 查询审计日志（分页 + 筛选）
#[command]
pub fn audit_query(
    state: State<'_, AppState>,
    query: AuditQuery,
) -> AppResult<AuditQueryResult> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let page = query.page.unwrap_or(1).max(1);
    let page_size = query.page_size.unwrap_or(20).clamp(1, 100);
    let offset = (page - 1) * page_size;

    // 构建 WHERE 子句
    let mut conditions = Vec::new();
    let mut param_values: Vec<Box<dyn rusqlite::types::ToSql>> = Vec::new();
    let mut idx = 1;

    if let Some(ref start) = query.start_date {
        conditions.push(format!("created_at >= ?{}", idx));
        param_values.push(Box::new(start.clone()));
        idx += 1;
    }
    if let Some(ref end) = query.end_date {
        conditions.push(format!("created_at <= ?{}", idx));
        param_values.push(Box::new(format!("{} 23:59:59", end)));
        idx += 1;
    }
    if let Some(ref action) = query.action {
        conditions.push(format!("action = ?{}", idx));
        param_values.push(Box::new(action.clone()));
        idx += 1;
    }
    if let Some(ref resource) = query.resource {
        conditions.push(format!("resource_type = ?{}", idx));
        param_values.push(Box::new(resource.clone()));
        idx += 1;
    }
    if let Some(ref keyword) = query.keyword {
        conditions.push(format!(
            "(user_name LIKE ?{0} OR details LIKE ?{0} OR resource_id LIKE ?{0})",
            idx
        ));
        param_values.push(Box::new(format!("%{}%", keyword)));
    }

    let where_clause = if conditions.is_empty() {
        String::new()
    } else {
        format!(" WHERE {}", conditions.join(" AND "))
    };

    // 查询总数
    let count_sql = format!("SELECT COUNT(*) FROM audit_logs{}", where_clause);
    let mut count_stmt = db.prepare(&count_sql).map_err(|e| AppError::Database(e.to_string()))?;
    let param_refs: Vec<&dyn rusqlite::types::ToSql> = param_values.iter().map(|p| p.as_ref()).collect();

    let total: i64 = count_stmt
        .query_row(param_refs.as_slice(), |row| row.get(0))
        .map_err(|e| AppError::Database(e.to_string()))?;

    // 查询数据
    let data_sql = format!(
        "SELECT id, user_id, user_name, action, resource_type, resource_id, details, ip_address, created_at
         FROM audit_logs{} ORDER BY created_at DESC LIMIT ?{} OFFSET ?{}",
        where_clause,
        idx,
        idx + 1
    );
    param_values.push(Box::new(page_size));
    param_values.push(Box::new(offset));

    let mut data_stmt = db.prepare(&data_sql).map_err(|e| AppError::Database(e.to_string()))?;
    let data_refs: Vec<&dyn rusqlite::types::ToSql> = param_values.iter().map(|p| p.as_ref()).collect();

    let items = data_stmt
        .query_map(data_refs.as_slice(), |row| {
            Ok(AuditLogEntry {
                id: row.get(0)?,
                user_id: row.get(1)?,
                username: row.get(2)?,
                action: row.get(3)?,
                resource: row.get(4)?,
                resource_id: row.get(5)?,
                resource_name: None,
                level: None,
                description: None,
                details: row.get(6)?,
                ip_address: row.get(7)?,
                timestamp: None,
                created_at: row.get(8)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?;

    let total_pages = (total as f64 / page_size as f64).ceil() as i64;

    Ok(AuditQueryResult {
        items,
        total,
        page,
        page_size,
        total_pages,
    })
}

/// 审计日志统计
#[command]
pub fn audit_stats(
    state: State<'_, AppState>,
    days: Option<i64>,
) -> AppResult<AuditStats> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;
    let now = chrono::Local::now();
    let today_prefix = now.format("%Y-%m-%d").to_string();

    // 时间范围过滤
    let date_filter = if let Some(d) = days {
        let cutoff = (now - chrono::Duration::days(d))
            .format("%Y-%m-%d %H:%M:%S")
            .to_string();
        format!(" WHERE created_at >= '{}'", cutoff)
    } else {
        String::new()
    };

    // 总数
    let total_count: i64 = db
        .query_row(
            &format!("SELECT COUNT(*) FROM audit_logs{}", date_filter),
            [],
            |row| row.get(0),
        )
        .unwrap_or(0);

    // 今日数量
    let today_count: i64 = db
        .query_row(
            &format!(
                "SELECT COUNT(*) FROM audit_logs{} AND created_at >= '{}'",
                if date_filter.is_empty() {
                    " WHERE".to_string()
                } else {
                    format!("{} AND", date_filter)
                },
                today_prefix
            ),
            [],
            |row| row.get(0),
        )
        .unwrap_or(0);

    // 操作分布
    let mut action_counts = std::collections::HashMap::new();
    {
        let sql = format!(
            "SELECT action, COUNT(*) FROM audit_logs{} GROUP BY action",
            date_filter
        );
        let mut stmt = db.prepare(&sql).map_err(|e| AppError::Database(e.to_string()))?;
        let rows = stmt
            .query_map([], |row| Ok((row.get::<_, String>(0)?, row.get::<_, i64>(1)?)))
            .map_err(|e| AppError::Database(e.to_string()))?;
        for row in rows {
            let (action, count) = row.map_err(|e| AppError::Database(e.to_string()))?;
            action_counts.insert(action, count);
        }
    }

    // 资源分布
    let mut resource_counts = std::collections::HashMap::new();
    {
        let sql = format!(
            "SELECT resource_type, COUNT(*) FROM audit_logs{} GROUP BY resource_type",
            date_filter
        );
        let mut stmt = db.prepare(&sql).map_err(|e| AppError::Database(e.to_string()))?;
        let rows = stmt
            .query_map([], |row| Ok((row.get::<_, String>(0)?, row.get::<_, i64>(1)?)))
            .map_err(|e| AppError::Database(e.to_string()))?;
        for row in rows {
            let (resource, count) = row.map_err(|e| AppError::Database(e.to_string()))?;
            resource_counts.insert(resource, count);
        }
    }

    // 活跃用户 TOP 10
    let mut top_users = Vec::new();
    {
        let sql = format!(
            "SELECT user_id, user_name, COUNT(*) FROM audit_logs{} GROUP BY user_id, user_name ORDER BY COUNT(*) DESC LIMIT 10",
            date_filter
        );
        let mut stmt = db.prepare(&sql).map_err(|e| AppError::Database(e.to_string()))?;
        let rows = stmt
            .query_map([], |row| {
                Ok(TopUser {
                    user_id: row.get::<_, String>(0)?,
                    username: row.get::<_, String>(1)?,
                    count: row.get::<_, i64>(2)?,
                })
            })
            .map_err(|e| AppError::Database(e.to_string()))?;
        for row in rows {
            top_users.push(row.map_err(|e| AppError::Database(e.to_string()))?);
        }
    }

    Ok(AuditStats {
        total_count,
        today_count,
        action_counts,
        resource_counts,
        top_users,
    })
}

/// 清理旧日志
#[command]
pub fn audit_clear(
    state: State<'_, AppState>,
    days_to_keep: i64,
) -> AppResult<i64> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let cutoff = (chrono::Local::now() - chrono::Duration::days(days_to_keep))
        .format("%Y-%m-%d %H:%M:%S")
        .to_string();

    let removed = db
        .execute(
            "DELETE FROM audit_logs WHERE created_at < ?1",
            params![cutoff],
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    Ok(removed as i64)
}
